src/coreComponents/linearAlgebra/interfaces/hypre/mgrStrategies/MultiphasePoromechanicsReservoirFVM.hpp

```diff
@@ -0,0 +1,156 @@
+ *   - dofLabel: numResLabels + numWellLabels + 1 = well density
+ *   - dofLabel: numResLabels + numWellLabels + 2 = well rate
+ *
+ * 4-level MGR reduction strategy
+ *   - 1st level: eliminate displacements (0,1,2)
+ *   - 2nd level: eliminate the well block
+ *   - 3nd level: eliminate the reservoir density associated with the volume constraint (numLabels - 1)
+ *   - 4nd level: eliminate the other reservoir densities
+ *   - The coarse grid is solved with BoomerAMG.
+ *
+ */
+class MultiphasePoromechanicsReservoirFVM : public MGRStrategyBase< 4 >
+{
+public:
+  /**
+   * @brief Constructor.
+   * @param numComponentsPerField array with number of components for each field
+   */
+  explicit MultiphasePoromechanicsReservoirFVM( arrayView1d< int const > const & numComponentsPerField )
+    : MGRStrategyBase( LvArray::integerConversion< HYPRE_Int >( numComponentsPerField[0] + numComponentsPerField[1] + numComponentsPerField[2] ) )
+  {
+    // Level 0: eliminate displacement degrees of freedom
+    m_labels[0].resize( m_numBlocks - 3 );
+    std::iota( m_labels[0].begin(), m_labels[0].end(), 3 );
+
+    HYPRE_Int const numResLabels = LvArray::integerConversion< HYPRE_Int >( numComponentsPerField[1] );
+
+    // Level 1: eliminate the well block
+    m_labels[1].resize( numResLabels );
+    std::iota( m_labels[1].begin(), m_labels[1].end(), 3 );
+    // Level 2: eliminate reservoir last density which corresponds to the volume constraint equation
+    m_labels[2].resize( numResLabels - 1 );
+    std::iota( m_labels[2].begin(), m_labels[2].end(), 3 );
+    // Level 3: eliminate the remaining reservoir densities
+    m_labels[3].push_back( 3 );
+
+    setupLabels();
+
+    // Level 0
+    m_levelFRelaxMethod[0]     = MGRFRelaxationMethod::amgVCycle;
+    m_levelInterpType[0]       = MGRInterpolationType::jacobi;
+    m_levelRestrictType[0]     = MGRRestrictionType::injection;
+    m_levelCoarseGridMethod[0] = MGRCoarseGridMethod::nonGalerkin;
+
+    // Level 1
+    m_levelFRelaxMethod[1]     = MGRFRelaxationMethod::singleLevel;
+    m_levelInterpType[1]       = MGRInterpolationType::blockJacobi;
+    m_levelRestrictType[1]     = MGRRestrictionType::injection;
+    m_levelCoarseGridMethod[1] = MGRCoarseGridMethod::galerkin;
+
+    // Level 2
+    m_levelFRelaxMethod[2]     = MGRFRelaxationMethod::singleLevel; //default, i.e. Jacobi
+    m_levelInterpType[2]       = MGRInterpolationType::jacobi;
+    m_levelRestrictType[2]     = MGRRestrictionType::injection;
+    m_levelCoarseGridMethod[2] = MGRCoarseGridMethod::galerkin;
+
+    // Level 3
+    m_levelFRelaxMethod[3]     = MGRFRelaxationMethod::singleLevel; //default, i.e. Jacobi
+    m_levelInterpType[3]       = MGRInterpolationType::injection;
+    m_levelRestrictType[3]     = MGRRestrictionType::injection;
+    m_levelCoarseGridMethod[3] = MGRCoarseGridMethod::cprLikeBlockDiag;
+
+    // Global smoother at each level, only do block-GS for the condensed system
+    m_levelSmoothType[3] = 1;
```

Comment:
I think at some point we were debating whether we wanted to have `HYPRE_ILU` as the default for this kind of global smoothing. Have you tested it?

Reply:
Yes, following your questions, I tested it for the following MGR recipes:
- `CompositionalMultiphaseFVM`
- `ThermalCompositionalMultiphaseFVM`
- `MultiphasePoromechanics`
- `CompositionalMultiphaseRevervoirFVM`
- `MultiphasePoromechanicsReservoirFVM`

It never hurts, and in most cases, it brings a large reduction in the number of linear iterations (especially for large time steps) and a reduction in run time (smaller than the reduction in the number of linear iterations, though).

So I am using this PR to make it the default smoother for the recipes mentioned above.